File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-prefer-pressable.ts
import { defineRule } from "../../utils/define-rule.js";
import type { Rule } from "../../utils/rule.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { getImportedName } from "../../utils/get-imported-name.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";

const TOUCHABLE_COMPONENTS = new Set([
  "TouchableOpacity",
  "TouchableHighlight",
  "TouchableWithoutFeedback",
  "TouchableNativeFeedback",
]);

const TOUCHABLE_SOURCES = new Set(["react-native", "react-native-gesture-handler"]);

// HACK: TouchableOpacity / TouchableHighlight / TouchableWithoutFeedback /
// TouchableNativeFeedback are legacy and feature-frozen. Pressable is the
// modern, more configurable, more accessible replacement that works the
// same on iOS, Android, and Fabric.
export const rnPreferPressable = defineRule<Rule>({
  id: "rn-prefer-pressable",
  title: "Touchable components instead of Pressable",
  tags: ["test-noise"],
  requires: ["react-native"],
  severity: "warn",
  recommendation:
    "Use `<Pressable>` because Touchable* components are frozen and lack Pressable's state-based feedback and accessibility behavior.",
  create: (context: RuleContext) => ({
    ImportDeclaration(node: EsTreeNodeOfType<"ImportDeclaration">) {
      const source = node.source?.value;
      if (typeof source !== "string" || !TOUCHABLE_SOURCES.has(source)) return;
      for (const specifier of node.specifiers ?? []) {
        if (!isNodeOfType(specifier, "ImportSpecifier")) continue;
        const importedName = getImportedName(specifier);
        if (!importedName || !TOUCHABLE_COMPONENTS.has(importedName)) continue;
        context.report({
          node: specifier,
          message: `Your users miss <Pressable>'s flexible press feedback when you use ${importedName}, which is old & frozen.`,
        });
      }
    },
  }),
});

File: native/oxlint/rules/rn-prefer-pressable.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{AstNode, context::LintContext, rule::Rule};

const TOUCHABLE_COMPONENTS: [&str; 4] = [
    "TouchableOpacity",
    "TouchableHighlight",
    "TouchableWithoutFeedback",
    "TouchableNativeFeedback",
];

#[derive(Debug, Default, Clone)]
pub struct RnPreferPressable;

declare_oxc_lint!(
    /// Prefer Pressable over frozen Touchable components.
    RnPreferPressable,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Prefer Pressable over frozen Touchable components.",
);

impl Rule for RnPreferPressable {
    fn should_run(&self, ctx: &crate::context::ContextHost) -> bool {
        !is_test_noise_file(ctx)
    }

    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::ImportDeclaration(import_declaration) = node.kind() else {
            return;
        };
        if !matches!(
            import_declaration.source.value.as_str(),
            "react-native" | "react-native-gesture-handler"
        ) {
            return;
        }
        for_each_value_import(import_declaration, |import_specifier| {
            let imported_name = import_specifier.imported.name();
            if !TOUCHABLE_COMPONENTS.contains(&imported_name.as_str()) {
                return;
            }
            ctx.diagnostic(
                OxcDiagnostic::warn(format!(
                    "Your users miss <Pressable>'s flexible press feedback when you use {imported_name}, which is old & frozen."
                ))
                .with_label(import_specifier.span),
            );
        });
    }
}
